client/meta_storage_client.go

```diff
@@ -0,0 +1,223 @@
+	options := &Op{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span = opentracing.StartSpan("pdclient.Put", opentracing.ChildOf(span.Context()))
+		defer span.Finish()
+	}
+	start := time.Now()
+	defer func() { cmdDurationPut.Observe(time.Since(start).Seconds()) }()
+
+	ctx, cancel := context.WithTimeout(ctx, c.option.timeout)
+	req := &meta_storagepb.PutRequest{
+		Key:    key,
+		Value:  value,
+		Lease:  options.lease,
+		PrevKv: options.prevKv,
+	}
+	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
+	resp, err := c.metaStorageClient().Put(ctx, req)
+	cancel()
+
+	if err = c.respForMetaStorageErr(cmdFailedDurationPut, start, err, resp.GetHeader()); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (c *client) Get(ctx context.Context, key []byte, opts ...OpOption) (*meta_storagepb.GetResponse, error) {
+	options := &Op{}
+	for _, opt := range opts {
+		opt(options)
+	}
+	if options.isOptsWithPrefix {
+		options.rangeEnd = getPrefix(key)
+	}
+
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span = opentracing.StartSpan("pdclient.Get", opentracing.ChildOf(span.Context()))
+		defer span.Finish()
+	}
+	start := time.Now()
+	defer func() { cmdDurationGet.Observe(time.Since(start).Seconds()) }()
+
+	ctx, cancel := context.WithTimeout(ctx, c.option.timeout)
+	req := &meta_storagepb.GetRequest{
+		Key:      key,
+		RangeEnd: options.rangeEnd,
+		Limit:    options.limit,
+		Revision: options.revision,
+	}
+	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
+	resp, err := c.metaStorageClient().Get(ctx, req)
+	cancel()
+
+	if err = c.respForMetaStorageErr(cmdFailedDurationGet, start, err, resp.GetHeader()); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (c *client) Watch(ctx context.Context, key []byte, opts ...OpOption) (chan []*meta_storagepb.Event, error) {
+	eventCh := make(chan []*meta_storagepb.Event, 100)
+	options := &Op{}
+	for _, opt := range opts {
+		opt(options)
+	}
+	if options.isOptsWithPrefix {
+		options.rangeEnd = getPrefix(key)
+	}
+
+	res, err := c.metaStorageClient().Watch(ctx, &meta_storagepb.WatchRequest{
+		Key:           key,
+		RangeEnd:      options.rangeEnd,
+		StartRevision: options.revision,
+		PrevKv:        options.prevKv,
+	})
+	if err != nil {
+		close(eventCh)
+		return nil, err
+	}
+	go func() {
+		defer func() {
+			close(eventCh)
+			if r := recover(); r != nil {
+				log.Error("[pd] panic in client `Watch`", zap.Any("error", r))
+				return
+			}
+		}()
+		for {
+			resp, err := res.Recv()
+			if err != nil {
+				return
+			}
+			select {
+			case <-ctx.Done():
+				return
+			case eventCh <- resp.GetEvents():
+			}
+		}
+	}()
+	return eventCh, err
+}
+
+func (c *client) respForMetaStorageErr(observer prometheus.Observer, start time.Time, err error, header *meta_storagepb.ResponseHeader) error {
```

Comment:
what's the difference with `client.respForErr`

Reply:
Almost the same, only the header is defined in a different proto, maybe we can use generic to do some refactoring later.